flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,811 @@
+          Preconditions.checkNotNull(
+              field,
+              "Missing required equality field column '%s' in table schema %s",
+              column,
+              table.schema());
+          equalityFieldSet.add(field.fieldId());
+        }
+
+        if (!equalityFieldSet.equals(table.schema().identifierFieldIds())) {
+          LOG.warn(
+              "The configured equality field column IDs {} are not matched with the schema identifier field IDs"
+                  + " {}, use job specified equality field columns as the equality fields by default.",
+              equalityFieldSet,
+              table.schema().identifierFieldIds());
+        }
+        equalityFieldIds = Lists.newArrayList(equalityFieldSet);
+      }
+      return equalityFieldIds;
+    }
+
+    public IcebergSink build() {
+
+      Preconditions.checkArgument(
+          inputCreator != null,
+          "Please use forRowData() or forMapperOutputType() to initialize the input DataStream.");
+      Preconditions.checkNotNull(tableLoader(), "Table loader shouldn't be null");
+
+      // Set the table if it is not yet set in the builder, so we can do the equalityId checks
+      table(checkAndGetTable(tableLoader(), table));
+
+      // Validate the equality fields and partition fields if we enable the upsert mode.
+      List<Integer> equalityFieldIds = checkAndGetEqualityFieldIds();
+
+      // Init the `flinkWriteConf` here, so we can do the checks
+      FlinkWriteConf flinkWriteConf = new FlinkWriteConf(table, writeOptions, readableConfig);
+      String branch = flinkWriteConf.branch();
+      boolean overwriteMode = flinkWriteConf.overwriteMode();
+      int workerPoolSize = flinkWriteConf.workerPoolSize();
+      if (flinkWriteConf.upsertMode()) {
+        Preconditions.checkState(
+            !flinkWriteConf.overwriteMode(),
+            "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+        Preconditions.checkState(
+            !equalityFieldIds.isEmpty(),
+            "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+        if (!table.spec().isUnpartitioned()) {
+          for (PartitionField partitionField : table.spec().fields()) {
+            Preconditions.checkState(
+                equalityFieldIds.contains(partitionField.sourceId()),
+                "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+                partitionField,
+                equalityFieldColumns);
+          }
+        }
+      }
+
+      Map<String, String> writeProperties =
+          writeProperties(table, flinkWriteConf.dataFileFormat(), flinkWriteConf);
+      RowType flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+      Duration tableRefreshInterval = flinkWriteConf.tableRefreshInterval();
+      SerializableSupplier<Table> tableSerializableSupplier;
```

Comment:
nit: like the simpler name `tableSupplier` from `FlinkSink`

Reply:
also is this really used/needed here. Previously, it was created and passed to writer operator. you can find the PR on more context

Reply:
`tableSupplier` will be passed to the `IcebergSinkWriter` inside the `icebergSink.createWriter` method

Reply:
`tableSupplier` is a stack local variable (not a class variable)

```
SerializableSupplier<Table> tableSupplier;
```

Reply:
`tableSupplier` variable is later on passed as a constructor argument below, when we instantiate the `IcebergSink`  ie:
```java
 return new IcebergSink(
          tableLoader,
          table,
          snapshotSummary,
          Optional.ofNullable(uidPrefix).orElse(""),
          writeProperties(table, flinkWriteConf.dataFileFormat(), flinkWriteConf),
          toFlinkRowType(table.schema(), tableSchema),
          tableSupplier,
          flinkWriteConf,
          equalityFieldIds,
          flinkWriteConf.branch(),
          overwriteMode);
    }
```